File: backend/internal/domain/question.go
package domain

import (
	"encoding/json"
	"errors"
	"time"
)

var ErrInvalidQuestionType = errors.New("invalid question type")

type QuestionType string

const (
	SingleChoice QuestionType = "single_choice"
	MultiChoice  QuestionType = "multi_choice"

	// Новые типы
	Consent   QuestionType = "consent"
	Email     QuestionType = "email"
	Rating    QuestionType = "rating"
	Date      QuestionType = "date"
	ShortText QuestionType = "short_text"
	LongText  QuestionType = "long_text"
	Number    QuestionType = "number"
)

// QuestionState – состояние временного вопроса относительно постоянной сущности.
type QuestionState string

const (
	QuestionStateActual  QuestionState = "ACTUAL"
	QuestionStateNew     QuestionState = "NEW"
	QuestionStateChanged QuestionState = "CHANGED"
	QuestionStateDeleted QuestionState = "DELETED"
)

// SurveyQuestion представляет вопрос опроса.
// Теперь добавляем поле QuestionOrder и удаляем Options.
type SurveyQuestion struct {
	ID            int             `json:"id" db:"id"`
	SurveyID      int             `json:"survey_id" db:"survey_id"`
	Label         string          `json:"label" db:"label"`
	Type          QuestionType    `json:"type" db:"type"`
	QuestionOrder int             `json:"question_order" db:"question_order"`
	ExtraParams   json.RawMessage `json:"extra_params" db:"extra_params"`
	Options       []Option        `json:"options,omitempty" db:"-"`
	CreatedAt     time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time       `json:"updated_at" db:"updated_at"`
}

// SurveyQuestionsTemp - временная таблица для вопросов.
type SurveyQuestionTemp struct {
	ID                 int             `json:"id" db:"id"`
	QuestionOriginalID *int            `json:"question_original_id" db:"question_original_id"`
	SurveyID           int             `json:"survey_id" db:"survey_id"`
	Label              string          `json:"label" db:"label"`
	Type               QuestionType    `json:"type" db:"type"`
	QuestionOrder      int             `json:"question_order" db:"question_order"`
	ExtraParams        json.RawMessage `json:"extra_params" db:"extra_params"`
	Options            []OptionTemp    `json:"options,omitempty" db:"-"`
	QuestionState      QuestionState   `json:"question_state" db:"question_state"`
	CreatedAt          time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt          time.Time       `json:"updated_at" db:"updated_at"`
}

File: frontend/survey-frontend/src/types/question.ts
// src/types/question.ts
import { Option } from "./option";
/** Тип вопроса */
export enum QuestionType {
    SingleChoice = "single_choice",
    MultiChoice  = "multi_choice",
    Consent      = "consent",
    Email        = "email",
    Rating       = "rating",
    Date         = "date",
    ShortText    = "short_text",
    LongText     = "long_text",
    Number       = "number",
  }
  
  /** Временные и постоянные состояния вопроса */
  export enum QuestionState {
    Actual  = "ACTUAL",
    New     = "NEW",
    Changed = "CHANGED",
    Deleted = "DELETED",
  }
  

  
  /** Полная структура вопроса с опциями */
  export interface SurveyQuestion {
    id: number;
    question_original_id: number;
    survey_id: number;
    label: string;
    type: QuestionType;
    question_order: number;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    extra_params: Record<string, any>;
    options?: Option[];
    question_state: QuestionState;
    created_at: string; // ISO
    updated_at: string; // ISO
  }
  
  /** Ответ сервера на создание/обновление одного вопроса */
  export interface QuestionResponse {
    question: SurveyQuestion;
  }
  
  /** Тело создания вопроса */
  export interface CreateQuestionRequest {
    type: QuestionType;
  }
  
  /** Тело обновления только метки */
  export interface UpdateQuestionLabelRequest {
    label: string;
  }
  
  /** Тело обновления порядка */
  export interface UpdateQuestionOrderRequest {
    new_order: number;
  }
  
  /** Тело обновления extra_params */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  export type UpdateQuestionExtraParamsRequest = Record<string, any>;
  